### ISSUE TITLE
Incorrect CS8825 `Return value must be non-null because parameter is non-null` in async methods

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#49077]
**Version Used**: 
```
Branch master (29 Oct 2020)
Latest commit ca87391 by David Wengier:
Don't crash when adding a parameter to the end of an extension method (#49007)
```

**Steps to Reproduce**:

```cs
#nullable enable
using System.Threading.Tasks;
using System.Diagnostics.CodeAnalysis;

string notNull = "";
var res = await M(notNull);
res.ToString(); // CS8602

[return: NotNullIfNotNull("s1")]
async Task<string?>? M(string? s1)
{
  s1 = "";
  return null; // CS8825
}
```

**Expected Behavior**:
No warnings in `M`

**Actual Behavior**:
`CS8825` is reported for `return null;` in `M` even though the null return is absolutely valid in `async Task<string?>` functions

**Notes**
The `[return: NotNullIfNotNull]` annotation produce warnings in async/interator methods because they can't return null values. Note that it doesn't affect nullability of the awaited expression on the call site i.e. `res.ToString()` report a warning, so the warning in `M` is meaningless.
Note that it still might make sense to annotate an async method with such annotation e.g. if you plan to override it with a non-async version that should support this contract.

FYI @RikkiGibson

**Comment by jcouv** [comment]
This is by-design.
We don't keep track of the initial value of the parameter, we only keep track of the value of the parameter. If we tried to keep track of the two separately, we'd run into other problems: a null test for one would be different than a null test for the other.

Note: the problem is not specific to async code:
```
#nullable enable
using System.Diagnostics.CodeAnalysis;

string notNull = "";
var res = M(notNull);
res.ToString(); // CS8602

[return: NotNullIfNotNull("s1")]
string? M(string? s1)
{
  s1 = "";
  return null; // CS8825
}
```

**Comment by TessenR** [comment]
@jcouv I think you misread the example. It's not about the actual state of the parameter but about the fact that the reutrned `null` value is the `string?` in `Task<string?>?` which is allowed to be nullable while the annotation `[NotNullIfNotNull("s")]` annotated the whole `Task<string?>?` which cannot be nullable in this method regardless of the returned value

**Comment by jcouv** [comment]
@TessenR Thanks for the clarification. I'd misunderstood indeed.
The question is whether the `[NotNullIfNotNull(...)]` applies to promised *values* or promises themselves. Right now, there is an inconsistency, as the caller interprets the attributes to apply to the promise but the implementer enforces the attribute on the returned *value*.
Since the caller cannot differentiate between `async` and regular methods, we must pick one interpretation (like that the attribute applies to the promise) and stick to it.
